sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/SupportsNamedArguments.scala

```diff
@@ -0,0 +1,204 @@
+   * If that object extends the trait [[SupportsNamedArguments]], then we rearrange
+   * the expressions in the order specified by the object.
+   *
+   * It is here we resubstitute [[Unevaluable]] [[NamedArgumentExpression]]s with
+   * normal expressions. This method will produce an positional argument list which
+   * is equivalent to the original argumnet list, except the expressions are now
+   * fit for consumption by [[ResolveFunctions]]
+   *
+   * @param expressions The list of positional and named argument expressions
+   * @tparam T The actual expression class.
+   * @return positional argument list
+   */
+  final def getRearrangedExpressions[T <: Expression : ClassTag](
+      expressions: Seq[Expression], functionName: String): Seq[Expression] = {
+
+    if (!expressions.exists(_.isInstanceOf[NamedArgumentExpression])) {
+      return expressions
+    }
+
+    import scala.reflect.runtime.currentMirror
+
+    // This code heavily utilizes Scala reflection which is unfamiliar to most developers.
+    // Here are the steps of this function:
+    // 1. Obtain the module symbol for the companion object of the function expression.
+    // 2. Obtain the module class symbol that represents the companion object.
+    // 3. Check if the base classes of the module class symbol contains SupportsNamedArguments.
+    //    This checks if the companion object is an implementor of SupportsNamedArguments.
+    // 4. Check if the module class symbol is a top level object. Reflection is unable to
+    //    obtain a companion object instance if it is member of some enclosing class unless
+    //    instance of said enclosing class is provided which we do not have.
+    // 5. Use reflection to obtain instance of companion object and perform immediate cast to
+    //    SupportsNamedArguments as it is already verified the cast is safe.
+    // 6. Obtain function signature and rearrange expression according to the given signature.
+    val runtimeClass = scala.reflect.classTag[T].runtimeClass
+    val targetModuleSymbol = currentMirror.classSymbol(runtimeClass).companion
+    val parentClass = scala.reflect.classTag[SupportsNamedArguments].runtimeClass
+    val parentSymbol = currentMirror.classSymbol(parentClass)
+
+    targetModuleSymbol match {
+      case scala.reflect.runtime.universe.NoSymbol =>
+        throw QueryCompilationErrors.namedArgumentsNotSupported(functionName)
+      case _ =>
+        val moduleClassSymbol = targetModuleSymbol.asModule.moduleClass.asClass
+        if (moduleClassSymbol.baseClasses.contains(parentSymbol)) {
+          if (currentMirror.runtimeClass(moduleClassSymbol).getEnclosingClass != null) {
+            throw QueryCompilationErrors.cannotObtainCompanionObjectInstance(functionName)
+          }
+          val instance = currentMirror.reflectModule(targetModuleSymbol.asModule)
+            .instance.asInstanceOf[SupportsNamedArguments]
```

Comment:
This is safe since earlier we do the check  ```moduleClassSymbol.baseClasses.contains(parentSymbol)``` . This line checks if any of the classes we inherit includes the SupportNamedArguments trait.